clients/vscode-hlasmplugin/src/hlasmDownloadCommands.ts

```diff
@@ -0,0 +1,695 @@
+
+        let wait = 0;
+        let result = { failed: new Array<JobDetail>(), total: 0 };
+        while (jobs.some(x => !x.downloaded)) {
+            checkCancel();
+
+            const list = (await client.list()).map(x => {
+                const j = jobsMap[x.jobname + "." + x.id];
+                return { fileInfo: x, job: j && !j.downloaded ? j : null };
+            }).filter(x => !!x.job);
+
+            for (const l of list) {
+                const job = l.job!;
+                job.unpacking = (await downloadJobAndProcess(client, l.fileInfo, job, progress, io)).unpacker
+                    .then(_ => { result.total++; })
+                    .catch(_ => { result.total++; result.failed.push(job.details); });
+            }
+
+            if (list.length === 0) {
+                if (wait < 30)
+                    wait += 1;
+                await new Promise((resolve) => setTimeout(resolve, wait * 1000))
+            }
+            else
+                wait = 0;
+        }
+
+        await Promise.all(jobs.map(x => x.unpacking));
+
+        return result;
+    }
+    finally {
+        client.dispose();
+    }
+}
+
+function askUser(prompt: string, password: boolean, defaultValue: string = ''): Promise<string> {
+    const input = vscode.window.createInputBox();
+    return new Promise<string>((resolve, reject) => {
+        input.prompt = prompt;
+        input.password = password;
+        input.value = defaultValue || '';
+        input.onDidHide(() => reject("Action was cancelled"));
+        input.onDidAccept(() => resolve(input.value));
+        input.show();
+    }).finally(() => { input.dispose(); });
+}
+
+function pickUser<T>(title: string, options: { label: string, value: T }[]): Promise<T> {
+    const input = vscode.window.createQuickPick();
+    return new Promise<T>((resolve, reject) => {
+        input.title = title;
+        input.items = options.map(x => { return { label: x.label }; });
+        input.canSelectMany = false;
+        input.onDidHide(() => reject("Action was cancelled"));
+        input.onDidAccept(() => resolve(options.find(x => x.label === input.selectedItems[0].label)!.value));
+        input.show();
+    }).finally(() => { input.dispose(); });
+}
+
+// because VSCode does not expose the service???
+export function replaceVariables(obj: any, resolver: (configKey: string) => (string | undefined), worksapceUri: vscode.Uri): any {
+    if (typeof obj === 'object') {
+        for (const x in obj)
+            obj[x] = replaceVariables(obj[x], resolver, worksapceUri);
+    }
+    else if (typeof obj === 'string') {
+        while (true) {
+            const match = /\$\{config:([^}]+)\}|\$\{(workspaceFolder)\}/.exec(obj);
+            if (!match) break;
+
+            const replacement = match[1] ? '' + resolver(match[1]) : worksapceUri;
+            obj = obj.slice(0, match.index) + replacement + obj.slice(match.index + match[0].length);
+        }
+    }
+    return obj;
+}
```

Comment:
To avoid the use of _any_ one can use _unknown_ and one type assertion like this
```suggestion
export function replaceVariables<T extends unknown>(
    obj: T,
    resolver: (configKey: string) => (string | undefined),
    worksapceUri: vscode.Uri): T {
    if (typeof obj === 'object') {
        for (const x in obj)
            obj[x] = replaceVariables(obj[x], resolver, worksapceUri);
        return obj;
    }
    else if (typeof obj === 'string') {
        let str: string = obj;
        while (true) {
            const match = /\$\{config:([^}]+)\}|\$\{(workspaceFolder)\}/.exec(str);
            if (!match) break;

            const replacement = match[1] ? '' + resolver(match[1]) : worksapceUri;
            str = str.slice(0, match.index) + replacement + str.slice(match.index + match[0].length);
        }
        return str as T;
    } else {
        return obj;
    }
}
}
```